// @webstudio-is/sdk-components-react-radix/lib/accordion.ws.js
import { AccordionIcon as t, ItemIcon as e, HeaderIcon as n, TriggerIcon as r, ContentIcon as c } from "@webstudio-is/icons/svg";
import { defaultStates as i } from "@webstudio-is/sdk";
import { h3 as a, button as p, div as s } from "@webstudio-is/sdk/normalize.css";
import { buttonReset as d } from "./shared/preset-styles.js";
import { propsAccordion as m, propsAccordionItem as l, propsAccordionHeader as A, propsAccordionTrigger as I, propsAccordionContent as g } from "./__generated__/accordion.props.js";
const o = {
  div: s
}, f = {
  type: "container",
  icon: t,
  presetStyle: o,
  constraints: [
    {
      relation: "descendant",
      component: { $eq: "AccordionItem" }
    }
  ]
}, v = {
  type: "container",
  label: "Item",
  icon: e,
  constraints: [
    {
      relation: "ancestor",
      component: { $eq: "Accordion" }
    },
    {
      relation: "descendant",
      component: { $eq: "AccordionHeader" }
    },
    {
      relation: "descendant",
      component: { $eq: "AccordionContent" }
    }
  ],
  indexWithinAncestor: "Accordion",
  presetStyle: o
}, H = {
  type: "container",
  label: "Item Header",
  icon: n,
  constraints: [
    {
      relation: "ancestor",
      component: { $eq: "AccordionItem" }
    },
    {
      relation: "descendant",
      component: { $eq: "AccordionTrigger" }
    }
  ],
  presetStyle: {
    h3: [
      ...a,
      {
        property: "margin-top",
        value: { type: "unit", unit: "px", value: 0 }
      },
      {
        property: "margin-bottom",
        value: { type: "unit", unit: "px", value: 0 }
      }
    ]
  }
}, C = {
  type: "container",
  label: "Item Trigger",
  icon: r,
  constraints: {
    relation: "ancestor",
    component: { $eq: "AccordionHeader" }
  },
  states: [
    ...i,
    {
      category: "component-states",
      label: "Open",
      selector: "[data-state=open]"
    }
  ],
  presetStyle: {
    button: [p, d].flat()
  }
}, T = {
  type: "container",
  label: "Item Content",
  icon: c,
  constraints: {
    relation: "ancestor",
    component: { $eq: "AccordionItem" }
  },
  presetStyle: o
}, M = {
  props: m,
  initialProps: ["value", "collapsible"]
}, x = {
  props: l,
  initialProps: ["value"]
}, S = {
  props: A
}, h = {
  props: I
}, P = {
  props: g
};
export {
  f as metaAccordion,
  T as metaAccordionContent,
  H as metaAccordionHeader,
  v as metaAccordionItem,
  C as metaAccordionTrigger,
  M as propsMetaAccordion,
  P as propsMetaAccordionContent,
  S as propsMetaAccordionHeader,
  x as propsMetaAccordionItem,
  h as propsMetaAccordionTrigger
};
