// @webstudio-is/sdk-components-react-radix/lib/checkbox.ws.js
import { CheckboxCheckedIcon as t, TriggerIcon as o } from "@webstudio-is/icons/svg";
import { defaultStates as e } from "@webstudio-is/sdk";
import { button as c, span as n } from "@webstudio-is/sdk/normalize.css";
import { buttonReset as a } from "./shared/preset-styles.js";
import { propsCheckbox as r, propsCheckboxIndicator as s } from "./__generated__/checkbox.props.js";
const k = {
  type: "container",
  constraints: {
    relation: "descendant",
    component: { $eq: "CheckboxIndicator" }
  },
  icon: t,
  states: [
    ...e,
    {
      label: "Checked",
      selector: "[data-state=checked]",
      category: "component-states"
    },
    {
      label: "Unchecked",
      selector: "[data-state=unchecked]",
      category: "component-states"
    }
  ],
  presetStyle: {
    button: [c, a].flat()
  }
}, l = {
  type: "container",
  constraints: {
    relation: "ancestor",
    component: { $eq: "Checkbox" }
  },
  icon: o,
  states: e,
  presetStyle: {
    span: n
  }
}, b = {
  props: r,
  initialProps: ["id", "className", "name", "value", "required", "checked"]
}, C = {
  props: s
};
export {
  k as metaCheckbox,
  l as metaCheckboxIndicator,
  b as propsMetaCheckbox,
  C as propsMetaCheckboxIndicator
};
