// @webstudio-is/sdk-components-react-radix/lib/collapsible.ws.js
import { CollapsibleIcon as e, TriggerIcon as n, ContentIcon as t } from "@webstudio-is/icons/svg";
import { div as o } from "@webstudio-is/sdk/normalize.css";
import { propsCollapsible as l, propsCollapsibleTrigger as p, propsCollapsibleContent as r } from "./__generated__/collapsible.props.js";
const c = {
  type: "container",
  constraints: [
    {
      relation: "descendant",
      component: { $eq: "CollapsibleTrigger" }
    },
    {
      relation: "descendant",
      component: { $eq: "CollapsibleContent" }
    }
  ],
  presetStyle: {
    div: o
  },
  icon: e
}, C = {
  type: "container",
  icon: n,
  constraints: {
    relation: "ancestor",
    component: { $eq: "Collapsible" }
  }
}, b = {
  type: "container",
  presetStyle: {
    div: o
  },
  icon: t,
  constraints: {
    relation: "ancestor",
    component: { $eq: "Collapsible" }
  }
}, m = {
  props: {
    ...l,
    onOpenChange: {
      type: "action",
      control: "action",
      required: !1
    }
  },
  initialProps: ["open", "onOpenChange"]
}, g = {
  props: p
}, d = {
  props: r
};
export {
  c as metaCollapsible,
  b as metaCollapsibleContent,
  C as metaCollapsibleTrigger,
  m as propsMetaCollapsible,
  d as propsMetaCollapsibleContent,
  g as propsMetaCollapsibleTrigger
};
