// @webstudio-is/sdk-components-react-radix/lib/dialog.ws.js
import { TriggerIcon as t, OverlayIcon as n, ContentIcon as e, HeadingIcon as i, TextIcon as a, ButtonElementIcon as r, DialogIcon as s } from "@webstudio-is/icons/svg";
import { defaultStates as p } from "@webstudio-is/sdk";
import { div as o, h2 as l, p as c, button as g } from "@webstudio-is/sdk/normalize.css";
import { propsDialog as D, propsDialogTrigger as m, propsDialogContent as y, propsDialogOverlay as C, propsDialogClose as d, propsDialogTitle as q, propsDialogDescription as $ } from "./__generated__/dialog.props.js";
import { buttonReset as T } from "./shared/preset-styles.js";
const O = {
  type: "container",
  icon: t,
  constraints: {
    relation: "ancestor",
    component: { $eq: "Dialog" }
  }
}, P = {
  type: "container",
  presetStyle: {
    div: o
  },
  icon: n,
  constraints: [
    {
      relation: "ancestor",
      component: { $eq: "Dialog" }
    },
    {
      relation: "descendant",
      component: { $eq: "DialogContent" }
    }
  ]
}, S = {
  type: "container",
  presetStyle: {
    div: o
  },
  icon: e,
  constraints: [
    {
      relation: "ancestor",
      component: { $eq: "DialogOverlay" }
    },
    // often deleted by users
    // though radix starts throwing warnings in console
    /*
    {
      relation: "descendant",
      component: { $eq: "DialogTitle" },
    },
    {
      relation: "descendant",
      component: { $eq: "DialogDescription" },
    },
    */
    {
      relation: "descendant",
      component: { $eq: "DialogClose" }
    }
  ]
}, b = {
  type: "container",
  presetStyle: {
    h2: l
  },
  icon: i,
  constraints: {
    relation: "ancestor",
    component: { $eq: "DialogContent" }
  }
}, x = {
  type: "container",
  presetStyle: {
    p: c
  },
  icon: a,
  constraints: {
    relation: "ancestor",
    component: { $eq: "DialogContent" }
  }
}, B = {
  type: "container",
  presetStyle: {
    button: [T, g].flat()
  },
  states: p,
  icon: r,
  label: "Close Button",
  constraints: {
    relation: "ancestor",
    component: { $eq: "DialogContent" }
  }
}, h = {
  type: "container",
  icon: s,
  constraints: [
    {
      relation: "descendant",
      component: { $eq: "DialogTrigger" }
    },
    {
      relation: "descendant",
      component: { $eq: "DialogOverlay" }
    }
  ]
}, E = {
  props: D,
  initialProps: []
}, H = {
  props: m
}, R = {
  props: y,
  initialProps: []
}, j = {
  props: C,
  initialProps: []
}, k = {
  props: d,
  initialProps: []
}, w = {
  props: q,
  initialProps: []
}, z = {
  props: $,
  initialProps: []
};
export {
  h as metaDialog,
  B as metaDialogClose,
  S as metaDialogContent,
  x as metaDialogDescription,
  P as metaDialogOverlay,
  b as metaDialogTitle,
  O as metaDialogTrigger,
  E as propsMetaDialog,
  k as propsMetaDialogClose,
  R as propsMetaDialogContent,
  z as propsMetaDialogDescription,
  j as propsMetaDialogOverlay,
  w as propsMetaDialogTitle,
  H as propsMetaDialogTrigger
};
