// @webstudio-is/sdk-components-react-radix/lib/navigation-menu.ws.js
import { NavigationMenuIcon as t, ListIcon as e, ListItemIcon as o, TriggerIcon as i, ContentIcon as a, BoxIcon as r, ViewportIcon as p } from "@webstudio-is/icons/svg";
import { div as n } from "@webstudio-is/sdk/normalize.css";
import { propsNavigationMenu as s, propsNavigationMenuItem as c, propsNavigationMenuTrigger as M, propsNavigationMenuContent as g, propsNavigationMenuLink as u, propsNavigationMenuList as v, propsNavigationMenuViewport as N } from "./__generated__/navigation-menu.props.js";
const y = {
  type: "container",
  icon: t,
  presetStyle: {
    div: n
  },
  constraints: [
    {
      relation: "descendant",
      component: { $eq: "NavigationMenuList" }
    },
    {
      relation: "descendant",
      component: { $eq: "NavigationMenuViewport" }
    }
  ]
}, L = {
  type: "container",
  icon: e,
  constraints: [
    {
      relation: "ancestor",
      component: { $eq: "NavigationMenu" }
    },
    {
      relation: "descendant",
      component: { $eq: "NavigationMenuItem" }
    }
  ],
  presetStyle: {
    div: n
  },
  label: "Menu List"
}, $ = {
  type: "container",
  icon: o,
  constraints: {
    relation: "ancestor",
    component: { $eq: "NavigationMenuList" }
  },
  presetStyle: {
    div: n
  },
  indexWithinAncestor: "NavigationMenu",
  label: "Menu Item"
}, d = {
  type: "container",
  icon: i,
  constraints: {
    relation: "ancestor",
    component: { $eq: "NavigationMenuItem" }
  },
  label: "Menu Trigger"
}, q = {
  type: "container",
  icon: a,
  constraints: {
    relation: "ancestor",
    component: { $eq: "NavigationMenuItem" }
  },
  indexWithinAncestor: "NavigationMenu",
  presetStyle: {
    div: n
  },
  label: "Menu Content"
}, b = {
  type: "container",
  icon: r,
  constraints: [
    {
      relation: "ancestor",
      component: { $eq: "NavigationMenu" }
    },
    {
      relation: "ancestor",
      component: { $in: ["NavigationMenuContent", "NavigationMenuItem"] }
    }
  ],
  label: "Accessible Link Wrapper"
}, w = {
  type: "container",
  icon: p,
  constraints: {
    relation: "ancestor",
    component: { $eq: "NavigationMenu" }
  },
  presetStyle: {
    div: n
  },
  label: "Menu Viewport"
}, C = {
  props: s
}, V = {
  props: c
}, S = {
  props: M
}, T = {
  props: g
}, k = {
  props: u
}, x = {
  props: v
}, f = {
  props: N
};
export {
  y as metaNavigationMenu,
  q as metaNavigationMenuContent,
  $ as metaNavigationMenuItem,
  b as metaNavigationMenuLink,
  L as metaNavigationMenuList,
  d as metaNavigationMenuTrigger,
  w as metaNavigationMenuViewport,
  C as propsMetaNavigationMenu,
  T as propsMetaNavigationMenuContent,
  V as propsMetaNavigationMenuItem,
  k as propsMetaNavigationMenuLink,
  x as propsMetaNavigationMenuList,
  S as propsMetaNavigationMenuTrigger,
  f as propsMetaNavigationMenuViewport
};
