// @webstudio-is/sdk-components-react-radix/lib/popover.ws.js
import { TriggerIcon as o, ContentIcon as e, PopoverIcon as n } from "@webstudio-is/icons/svg";
import { div as t } from "@webstudio-is/sdk/normalize.css";
import { propsPopover as r, propsPopoverTrigger as p, propsPopoverContent as i } from "./__generated__/popover.props.js";
const P = {
  type: "container",
  icon: o,
  constraints: {
    relation: "ancestor",
    component: { $eq: "Popover" }
  }
}, v = {
  type: "container",
  presetStyle: {
    div: t
  },
  icon: e,
  constraints: {
    relation: "ancestor",
    component: { $eq: "Popover" }
  }
}, m = {
  type: "container",
  icon: n,
  constraints: [
    {
      relation: "descendant",
      component: { $eq: "PopoverTrigger" }
    },
    {
      relation: "descendant",
      component: { $eq: "PopoverContent" }
    }
  ]
}, g = {
  props: r,
  initialProps: ["open"]
}, l = {
  props: p
}, d = {
  props: i,
  initialProps: ["side", "sideOffset", "align", "alignOffset"]
};
export {
  m as metaPopover,
  v as metaPopoverContent,
  P as metaPopoverTrigger,
  g as propsMetaPopover,
  d as propsMetaPopoverContent,
  l as propsMetaPopoverTrigger
};
