// @webstudio-is/sdk-components-react-radix/lib/radio-group.ws.js
import { RadioGroupIcon as t, ItemIcon as e, TriggerIcon as a } from "@webstudio-is/icons/svg";
import { defaultStates as o } from "@webstudio-is/sdk";
import { div as r, button as n, span as p } from "@webstudio-is/sdk/normalize.css";
import { buttonReset as s } from "./shared/preset-styles.js";
import { propsRadioGroup as i, propsRadioGroupItem as c, propsRadioGroupIndicator as d } from "./__generated__/radio-group.props.js";
const I = {
  type: "container",
  constraints: {
    relation: "descendant",
    component: { $eq: "RadioGroupItem" }
  },
  icon: t,
  states: [
    ...o,
    {
      label: "Checked",
      selector: "[data-state=checked]",
      category: "component-states"
    },
    {
      label: "Unchecked",
      selector: "[data-state=unchecked]",
      category: "component-states"
    }
  ],
  presetStyle: {
    div: r
  }
}, y = {
  type: "container",
  constraints: [
    {
      relation: "ancestor",
      component: { $eq: "RadioGroup" }
    },
    {
      relation: "descendant",
      component: { $eq: "RadioGroupIndicator" }
    }
  ],
  icon: e,
  states: o,
  presetStyle: {
    button: [n, s].flat()
  }
}, f = {
  type: "container",
  icon: a,
  constraints: {
    relation: "ancestor",
    component: { $eq: "RadioGroupItem" }
  },
  states: o,
  presetStyle: {
    span: p
  }
}, b = {
  props: i,
  initialProps: ["id", "className", "name", "value", "required"]
}, q = {
  props: c,
  initialProps: ["value"]
}, g = {
  props: d
};
export {
  I as metaRadioGroup,
  f as metaRadioGroupIndicator,
  y as metaRadioGroupItem,
  b as propsMetaRadioGroup,
  g as propsMetaRadioGroupIndicator,
  q as propsMetaRadioGroupItem
};
