// @webstudio-is/sdk-components-react-radix/lib/select.ws.js
import { SelectIcon as o, TriggerIcon as n, FormTextFieldIcon as c, ContentIcon as r, ViewportIcon as p, ItemIcon as a, CheckMarkIcon as s, TextIcon as l } from "@webstudio-is/icons/svg";
import { button as i, span as e, div as t } from "@webstudio-is/sdk/normalize.css";
import { propsSelect as m, propsSelectTrigger as S, propsSelectValue as I, propsSelectContent as d, propsSelectViewport as q, propsSelectItem as y, propsSelectItemIndicator as $, propsSelectItemText as T } from "./__generated__/select.props.js";
const M = {
  type: "container",
  icon: o,
  constraints: [
    {
      relation: "descendant",
      component: { $eq: "SelectTrigger" }
    },
    {
      relation: "descendant",
      component: { $eq: "SelectContent" }
    }
  ]
}, x = {
  type: "container",
  icon: n,
  presetStyle: {
    button: i
  },
  constraints: [
    {
      relation: "ancestor",
      component: { $eq: "Select" }
    },
    {
      relation: "descendant",
      component: { $eq: "SelectValue" }
    }
  ]
}, C = {
  type: "container",
  label: "Value",
  icon: c,
  presetStyle: {
    span: e
  },
  constraints: {
    relation: "ancestor",
    component: { $eq: "SelectTrigger" }
  }
}, w = {
  type: "container",
  icon: r,
  presetStyle: {
    div: t
  },
  constraints: [
    {
      relation: "ancestor",
      component: { $eq: "Select" }
    },
    {
      relation: "descendant",
      component: { $eq: "SelectViewport" }
    }
  ]
}, b = {
  type: "container",
  icon: p,
  presetStyle: {
    div: t
  },
  constraints: [
    {
      relation: "ancestor",
      component: { $eq: "SelectContent" }
    },
    {
      relation: "descendant",
      component: { $eq: "SelectItem" }
    }
  ]
}, f = {
  type: "container",
  icon: a,
  constraints: [
    {
      relation: "ancestor",
      component: { $eq: "SelectViewport" }
    },
    {
      relation: "descendant",
      component: { $eq: "SelectItemIndicator" }
    },
    {
      relation: "descendant",
      component: { $eq: "SelectItemText" }
    }
  ],
  presetStyle: {
    div: t
  }
}, v = {
  type: "container",
  label: "Indicator",
  icon: s,
  constraints: {
    relation: "ancestor",
    component: { $eq: "SelectItem" }
  },
  presetStyle: {
    span: e
  }
}, P = {
  type: "container",
  label: "Item Text",
  icon: l,
  constraints: {
    relation: "ancestor",
    component: { $eq: "SelectItem" }
  },
  presetStyle: {
    span: e
  }
}, h = {
  props: m,
  initialProps: ["name", "value", "open", "required"]
}, k = {
  props: S
}, F = {
  props: I,
  initialProps: ["placeholder"]
}, j = {
  props: d
}, z = {
  props: q
}, A = {
  props: y,
  initialProps: ["value"]
}, B = {
  props: $
}, D = {
  props: T
};
export {
  M as metaSelect,
  w as metaSelectContent,
  f as metaSelectItem,
  v as metaSelectItemIndicator,
  P as metaSelectItemText,
  x as metaSelectTrigger,
  C as metaSelectValue,
  b as metaSelectViewport,
  h as propsMetaSelect,
  j as propsMetaSelectContent,
  A as propsMetaSelectItem,
  B as propsMetaSelectItemIndicator,
  D as propsMetaSelectItemText,
  k as propsMetaSelectTrigger,
  F as propsMetaSelectValue,
  z as propsMetaSelectViewport
};
