// @webstudio-is/sdk-components-react-radix/lib/switch.ws.js
import { SwitchIcon as e, TriggerIcon as o } from "@webstudio-is/icons/svg";
import { defaultStates as t } from "@webstudio-is/sdk";
import { button as c, span as a } from "@webstudio-is/sdk/normalize.css";
import { buttonReset as s } from "./shared/preset-styles.js";
import { propsSwitch as n, propsSwitchThumb as r } from "./__generated__/switch.props.js";
const d = {
  type: "container",
  constraints: {
    relation: "descendant",
    component: { $eq: "SwitchThumb" }
  },
  icon: e,
  states: [
    ...t,
    {
      label: "Checked",
      selector: "[data-state=checked]",
      category: "component-states"
    },
    {
      label: "Unchecked",
      selector: "[data-state=unchecked]",
      category: "component-states"
    }
  ],
  presetStyle: {
    button: [c, s].flat()
  }
}, u = {
  type: "container",
  constraints: {
    relation: "ancestor",
    component: { $eq: "Switch" }
  },
  icon: o,
  states: [
    ...t,
    {
      label: "Checked",
      selector: "[data-state=checked]",
      category: "component-states"
    },
    {
      label: "Unchecked",
      selector: "[data-state=unchecked]",
      category: "component-states"
    }
  ],
  presetStyle: {
    span: a
  }
}, S = {
  props: n,
  initialProps: ["id", "className", "name", "value", "checked", "required"]
}, b = {
  props: r
};
export {
  d as metaSwitch,
  u as metaSwitchThumb,
  S as propsMetaSwitch,
  b as propsMetaSwitchThumb
};
