// @webstudio-is/sdk-components-react-radix/lib/tabs.js
import { jsx as i } from "react/jsx-runtime";
import { forwardRef as c, useCallback as m } from "react";
import { Root as f, List as b, Trigger as p, Content as T } from "@radix-ui/react-tabs";
import { getIndexWithinAncestorFromComponentProps as g, getClosestInstance as l } from "@webstudio-is/react-sdk/runtime";
import { useControllableState as d } from "@radix-ui/react-use-controllable-state";
import h from "await-interaction-response";
const S = c(({ defaultValue: e, ...t }, o) => {
  const [n, a] = d({
    prop: t.value,
    defaultProp: e,
    onChange: t.onValueChange
  }), s = m(
    async (u) => {
      await h(), a(u);
    },
    [a]
  );
  return /* @__PURE__ */ i(
    f,
    {
      ref: o,
      ...t,
      value: n,
      onValueChange: s
    }
  );
}), k = b, w = c(({ value: e, ...t }, o) => {
  const n = g(t);
  return /* @__PURE__ */ i(p, { ref: o, value: e ?? n ?? "", ...t });
}), y = c(({ value: e, ...t }, o) => {
  const n = g(t);
  return /* @__PURE__ */ i(T, { ref: o, value: e ?? n ?? "", ...t });
}), r = "@webstudio-is/sdk-components-react-radix", A = {
  onNavigatorSelect: (e, t) => {
    var o;
    for (const n of t.instancePath)
      if (n.component === `${r}:TabsContent` || n.component === `${r}:TabsTrigger`) {
        const a = l(
          t.instancePath,
          n,
          `${r}:Tabs`
        ), s = e.getPropValue(n, "value") ?? ((o = e.indexesWithinAncestors.get(n.id)) == null ? void 0 : o.toString());
        a && s && e.setMemoryProp(a, "value", s);
      }
  },
  onNavigatorUnselect: (e, t) => {
    var o;
    for (const n of t.instancePath)
      if (n.component === `${r}:TabsContent` || n.component === `${r}:TabsTrigger`) {
        const a = l(
          t.instancePath,
          n,
          `${r}:Tabs`
        ), s = e.getPropValue(n, "value") ?? ((o = e.indexesWithinAncestors.get(n.id)) == null ? void 0 : o.toString());
        a && s && e.setMemoryProp(a, "value", void 0);
      }
  }
};
export {
  S as Tabs,
  y as TabsContent,
  k as TabsList,
  w as TabsTrigger,
  A as hooksTabs
};
