// @webstudio-is/sdk-components-react-radix/lib/tabs.ws.js
import { TabsIcon as e, HeaderIcon as n, TriggerIcon as o, ContentIcon as s } from "@webstudio-is/icons/svg";
import { defaultStates as a } from "@webstudio-is/sdk";
import { button as r, div as c } from "@webstudio-is/sdk/normalize.css";
import { buttonReset as i } from "./shared/preset-styles.js";
import { propsTabs as p, propsTabsList as T, propsTabsTrigger as b, propsTabsContent as m } from "./__generated__/tabs.props.js";
const t = {
  div: c
}, y = {
  type: "container",
  icon: e,
  constraints: [
    {
      relation: "descendant",
      component: { $eq: "TabsTrigger" }
    },
    {
      relation: "descendant",
      component: { $eq: "TabsList" }
    },
    {
      relation: "descendant",
      component: { $eq: "TabsContent" }
    }
  ],
  presetStyle: t
}, $ = {
  type: "container",
  icon: n,
  constraints: {
    relation: "ancestor",
    component: { $eq: "Tabs" }
  },
  presetStyle: t
}, C = {
  type: "container",
  icon: o,
  constraints: [
    {
      relation: "ancestor",
      component: { $eq: "TabsList" }
    },
    {
      relation: "ancestor",
      component: { $neq: "TabsTrigger" }
    }
  ],
  indexWithinAncestor: "Tabs",
  label: "Tab Trigger",
  states: [
    ...a,
    {
      category: "component-states",
      label: "Active",
      selector: "[data-state=active]"
    }
  ],
  presetStyle: {
    button: [r, i].flat()
  }
}, L = {
  type: "container",
  label: "Tab Content",
  icon: s,
  constraints: {
    relation: "ancestor",
    component: { $eq: "Tabs" }
  },
  indexWithinAncestor: "Tabs",
  presetStyle: t
}, u = {
  props: p
}, I = {
  props: T
}, M = {
  props: b
}, v = {
  props: m
};
export {
  y as metaTabs,
  L as metaTabsContent,
  $ as metaTabsList,
  C as metaTabsTrigger,
  u as propsMetaTabs,
  v as propsMetaTabsContent,
  I as propsMetaTabsList,
  M as propsMetaTabsTrigger
};
