// @webstudio-is/sdk-components-react-radix/lib/tooltip.ws.js
import { TriggerIcon as o, ContentIcon as t, TooltipIcon as n } from "@webstudio-is/icons/svg";
import { div as e } from "@webstudio-is/sdk/normalize.css";
import { propsTooltip as i, propsTooltipTrigger as p, propsTooltipContent as r } from "./__generated__/tooltip.props.js";
const l = {
  type: "container",
  icon: o,
  constraints: {
    relation: "ancestor",
    component: { $eq: "Tooltip" }
  }
}, T = {
  type: "container",
  icon: t,
  presetStyle: {
    div: e
  },
  constraints: {
    relation: "ancestor",
    component: { $eq: "Tooltip" }
  }
}, m = {
  type: "container",
  constraints: [
    {
      relation: "descendant",
      component: { $eq: "TooltipTrigger" }
    },
    {
      relation: "descendant",
      component: { $eq: "TooltipContent" }
    }
  ],
  icon: n
}, g = {
  props: i,
  initialProps: ["open", "delayDuration", "disableHoverableContent"]
}, d = {
  props: p
}, f = {
  props: r,
  initialProps: ["side", "sideOffset", "align", "alignOffset"]
};
export {
  m as metaTooltip,
  T as metaTooltipContent,
  l as metaTooltipTrigger,
  g as propsMetaTooltip,
  f as propsMetaTooltipContent,
  d as propsMetaTooltipTrigger
};
