// @webstudio-is/sdk-components-react/lib/checkbox.ws.js
import { CheckboxCheckedIcon as e } from "@webstudio-is/icons/svg";
import { defaultStates as t } from "@webstudio-is/sdk";
import { checkbox as o } from "@webstudio-is/sdk/normalize.css";
import { props as l } from "./__generated__/checkbox.props.js";
const r = {
  input: [
    ...o,
    {
      property: "margin-right",
      value: { type: "unit", unit: "em", value: 0.5 }
    }
  ]
}, s = {
  constraints: {
    relation: "ancestor",
    component: { $nin: ["Button", "Link"] }
  },
  type: "control",
  icon: e,
  presetStyle: r,
  states: [
    ...t,
    { selector: ":checked", label: "Checked" },
    { selector: ":required", label: "Required" },
    { selector: ":optional", label: "Optional" },
    { selector: ":disabled", label: "Disabled" },
    { selector: ":enabled", label: "Enabled" },
    { selector: ":read-only", label: "Read Only" },
    { selector: ":read-write", label: "Read Write" }
  ]
}, p = {
  props: l,
  initialProps: ["id", "className", "name", "value", "required", "checked"]
};
export {
  s as meta,
  p as propsMeta
};
