// @webstudio-is/sdk-components-react/lib/vimeo.js
import { jsx as s, jsxs as D, Fragment as F } from "react/jsx-runtime";
import { colord as z } from "colord";
import { createContext as T, forwardRef as q, useState as v, useContext as G, useEffect as f } from "react";
import { ReactSdkContext as H } from "@webstudio-is/react-sdk/runtime";
const Y = (e) => {
  if (e.url === void 0)
    return;
  let t;
  try {
    const o = new URL(e.url);
    t = new URL(R), t.pathname = `/video${o.pathname}`;
  } catch {
  }
  if (t === void 0)
    return;
  const a = {
    showPortrait: "portrait",
    showByline: "byline",
    showTitle: "title",
    controlsColor: "color",
    showControls: "controls",
    interactiveParams: "interactive_params",
    backgroundMode: "background",
    doNotTrack: "dnt"
  };
  let r;
  for (r in e) {
    const o = e[r];
    if (r === "url" || o === void 0)
      continue;
    const i = a[r] ?? r;
    t.searchParams.append(i, o.toString());
  }
  if (t.searchParams.set("autoplay", "true"), typeof e.controlsColor == "string") {
    const o = z(e.controlsColor).toHex().replace("#", "");
    t.searchParams.set("color", o);
  }
  return e.showPortrait && t.searchParams.set("title", "true"), e.showByline && (t.searchParams.set("portrait", "true"), t.searchParams.set("title", "true")), t.toString();
}, h = (e) => {
  const t = document.createElement("link");
  t.rel = "preconnect", t.href = e, t.crossOrigin = "true", document.head.appendChild(t);
};
let b = !1;
const J = "https://f.vimeocdn.com", R = "https://player.vimeo.com", U = "https://i.vimeocdn.com", K = () => {
  b || window.matchMedia("(hover: none)").matches || (h(J), h(R), h(U), b = !0);
}, Q = (e) => {
  try {
    const a = new URL(e).pathname.split("/")[2];
    return a === "" || a == null ? void 0 : a;
  } catch {
  }
}, W = async (e) => {
  const a = `https://vimeo.com/api/v2/video/${Q(e)}.json`, o = (await (await fetch(a)).json())[0].thumbnail_large, i = o.substr(o.lastIndexOf("/") + 1).split("_")[0], n = new URL(U);
  return n.pathname = `/video/${i}.webp`, n.searchParams.append("mw", "1100"), n.searchParams.append("mh", "619"), n.searchParams.append("q", "70"), n;
}, X = () => /* @__PURE__ */ s(
  "div",
  {
    style: {
      display: "flex",
      width: "100%",
      height: "100%",
      alignItems: "center",
      justifyContent: "center",
      fontSize: "1.2em"
    },
    children: 'Open the "Settings" panel and paste a video URL, e.g. https://vimeo.com/831343124.'
  }
), Z = ({
  title: e,
  status: t,
  loading: a,
  videoUrl: r,
  previewImageUrl: o,
  autoplay: i,
  renderer: n,
  showPreview: d,
  onStatusChange: c,
  onPreviewImageUrlChange: l
}) => {
  const [p, u] = v(0);
  if (f(() => {
    i && n !== "canvas" && t === "initial" && c("loading");
  }, [i, t, n, c]), f(() => {
    n !== "canvas" && K();
  }, [n]), f(() => {
    if (r !== void 0) {
      if (d === !1) {
        l(void 0);
        return;
      }
      o === void 0 && W(r).then(l).catch(() => {
        console.error(`Could not load preview image for ${r}`);
      });
    }
  }, [l, d, r, o]), !(n === "canvas" || t === "initial"))
    return /* @__PURE__ */ s(
      "iframe",
      {
        title: e,
        src: r,
        loading: a,
        allow: "accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture;",
        allowFullScreen: !0,
        style: {
          position: "absolute",
          width: "100%",
          height: "100%",
          opacity: p,
          transition: "opacity 1s",
          border: "none"
        },
        onLoad: () => {
          c("ready"), u(1);
        }
      }
    );
}, tt = T({
  onInitPlayer: () => {
  },
  status: "initial"
}), et = q(
  ({
    url: e,
    loading: t = "lazy",
    autoplay: a = !1,
    autopause: r = !0,
    showByline: o = !1,
    showControls: i = !0,
    doNotTrack: n = !1,
    keyboard: d = !0,
    loop: c = !1,
    muted: l = !1,
    pip: p = !1,
    playsinline: u = !0,
    showPortrait: S = !0,
    quality: L = "auto",
    responsive: k = !0,
    speed: E = !1,
    showTitle: _ = !1,
    transparent: j = !0,
    showPreview: M = !1,
    autopip: N,
    controlsColor: O,
    interactiveParams: V,
    texttrack: $,
    children: A,
    ...y
  }, m) => {
    const [w, g] = v("initial"), [P, B] = v(), { renderer: C } = G(H), I = Y({
      url: e,
      autoplay: a,
      autopause: r,
      showControls: i,
      controlsColor: O,
      doNotTrack: n,
      interactiveParams: V,
      keyboard: d,
      loop: c,
      muted: l,
      pip: p,
      playsinline: u,
      quality: L,
      responsive: k,
      speed: E,
      texttrack: $,
      showTitle: _,
      transparent: j,
      showPortrait: S,
      autopip: N
    });
    return /* @__PURE__ */ s(
      tt.Provider,
      {
        value: {
          status: w,
          previewImageUrl: P,
          onInitPlayer() {
            C !== "canvas" && g("loading");
          }
        },
        children: /* @__PURE__ */ s(
          "div",
          {
            ...y,
            ref: (x) => {
              m !== null && (typeof m == "function" ? m(x) : m.current = x);
            },
            children: I === void 0 ? /* @__PURE__ */ s(X, {}) : /* @__PURE__ */ D(F, { children: [
              A,
              /* @__PURE__ */ s(
                Z,
                {
                  title: y.title,
                  autoplay: a,
                  videoUrl: I,
                  previewImageUrl: P,
                  loading: t,
                  showPreview: M,
                  renderer: C,
                  status: w,
                  onStatusChange: g,
                  onPreviewImageUrlChange: B
                }
              )
            ] })
          }
        )
      }
    );
  }
);
et.displayName = "Vimeo";
export {
  et as Vimeo,
  tt as VimeoContext
};
