// @webstudio-is/sdk-components-react/lib/youtube.js
import { jsx as u, jsxs as R, Fragment as _ } from "react/jsx-runtime";
import { forwardRef as L, useState as h, useContext as C, useEffect as y } from "react";
import { ReactSdkContext as I } from "@webstudio-is/react-sdk/runtime";
import { VimeoContext as E } from "./vimeo.js";
const O = "https://www.youtube-nocookie.com", x = "https://www.youtube.com", w = "https://img.youtube.com", k = (e) => {
  if (e)
    try {
      const t = new URL(e);
      return t.pathname === "/embed" ? void 0 : t.hostname === "youtu.be" ? t.pathname.slice(1) : t.searchParams.get("v") || t.pathname.split("/").pop();
    } catch {
      return e;
    }
}, S = (e, t) => {
  var n, i;
  const l = k(e.url), r = new URL(t);
  if (l)
    r.pathname = `/embed/${l}`;
  else if (e.url)
    try {
      const c = new URL(e.url);
      r.pathname = c.pathname, r.search = c.search;
    } catch {
    }
  const o = Object.keys(e), a = {};
  for (const c of o)
    switch (c) {
      case "autoplay":
        a.autoplay = e.autoplay ? "1" : "0", e.autoplay && e.muted === void 0 && (a.mute = "1");
        break;
      case "muted":
        a.mute = e.muted ? "1" : "0";
        break;
      case "showControls":
        a.controls = e.showControls ? "1" : "0";
        break;
      case "showRelatedVideos":
        a.rel = e.showRelatedVideos ? "1" : "0";
        break;
      case "keyboard":
        a.keyboard = e.keyboard ? "1" : "0";
        break;
      case "loop":
        a.loop = e.loop ? "1" : "0", e.loop && (e.playlist ?? "").trim() === "" && (a.playlist = l);
        break;
      case "inline":
        a.playsinline = e.inline ? "1" : "0";
        break;
      case "allowFullscreen":
        a.fs = e.allowFullscreen ? "1" : "0";
        break;
      case "captionLanguage":
        a.cc_lang_pref = e.captionLanguage;
        break;
      case "showCaptions":
        a.cc_load_policy = e.showCaptions ? "1" : "0";
        break;
      case "showAnnotations":
        a.iv_load_policy = e.showAnnotations ? "1" : "3";
        break;
      case "startTime":
        a.start = (n = e.startTime) == null ? void 0 : n.toString();
        break;
      case "endTime":
        a.end = (i = e.endTime) == null ? void 0 : i.toString();
        break;
      case "disableKeyboard":
        a.disablekb = e.disableKeyboard ? "1" : "0";
        break;
      case "language":
        a.hl = e.language;
        break;
      case "listId":
        a.list = e.listId;
        break;
      case "listType":
        a.listType = e.listType;
        break;
      case "color":
        a.color = e.color;
        break;
      case "origin":
        a.origin = e.origin;
        break;
      case "referrer":
        a.widget_referrer = e.referrer;
        break;
      case "playlist":
        a.playlist = e.playlist;
        break;
    }
  return Object.entries(a).forEach(([c, s]) => {
    s !== void 0 && r.searchParams.append(c, s.toString());
  }), r.toString();
}, p = (e) => {
  const t = document.createElement("link");
  t.rel = "preconnect", t.href = e, t.crossOrigin = "true", document.head.appendChild(t);
};
let g = !1;
const T = (e) => {
  if (!(g || window.matchMedia("(hover: none)").matches)) {
    try {
      const t = new URL(e);
      p(t.origin);
    } catch {
    }
    p(w), g = !0;
  }
}, j = (e) => new URL(`${w}/vi/${e}/maxresdefault.jpg`), A = () => /* @__PURE__ */ u("div", { className: "flex w-full h-full items-center justify-center text-lg", children: 'Open the "Settings" panel and paste a video URL, e.g. https://youtube.com/watch?v=dQw4w9WgXcQ' }), N = ({
  title: e,
  status: t,
  loading: l,
  videoUrl: r,
  previewImageUrl: o,
  autoplay: a,
  renderer: n,
  showPreview: i,
  onStatusChange: c,
  onPreviewImageUrlChange: s
}) => {
  const [m, b] = h(0);
  return y(() => {
    a && n !== "canvas" && t === "initial" && c("loading");
  }, [a, t, n, c]), y(() => {
    n !== "canvas" && T(r);
  }, [n, r]), y(() => {
    const d = k(r);
    if (!d || !i) {
      s(void 0);
      return;
    }
    o || s(j(d));
  }, [s, i, r, o]), n === "canvas" || t === "initial" ? null : /* @__PURE__ */ u(
    "iframe",
    {
      title: e,
      src: r,
      loading: l,
      allow: "accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture",
      allowFullScreen: !0,
      style: {
        position: "absolute",
        width: "100%",
        height: "100%",
        opacity: m,
        transition: "opacity 1s",
        border: "none"
      },
      onLoad: () => {
        c("ready"), b(1);
      }
    }
  );
}, P = L(
  ({
    url: e,
    loading: t = "lazy",
    autoplay: l,
    showPreview: r,
    children: o,
    privacyEnhancedMode: a,
    ...n
  }, i) => {
    const [c, s] = h("initial"), [m, b] = h(), { renderer: d } = C(I), v = a ?? !0 ? O : x, f = S(
      {
        ...n,
        url: e,
        autoplay: !0
      },
      v
    );
    return /* @__PURE__ */ u(
      E.Provider,
      {
        value: {
          status: c,
          previewImageUrl: m,
          onInitPlayer() {
            d !== "canvas" && s("loading");
          }
        },
        children: /* @__PURE__ */ u("div", { ...n, ref: i, children: f ? /* @__PURE__ */ R(_, { children: [
          o,
          /* @__PURE__ */ u(
            N,
            {
              title: n.title,
              autoplay: l,
              videoUrl: f,
              previewImageUrl: m,
              loading: t,
              showPreview: r,
              renderer: d,
              status: c,
              onStatusChange: s,
              onPreviewImageUrlChange: b
            }
          )
        ] }) : /* @__PURE__ */ u(A, {}) })
      }
    );
  }
);
P.displayName = "YouTube";
export {
  P as YouTube
};
